### ISSUE TITLE
Superfluous deep_copy() calls

### ISSUE THREAD

**Issue by jmmcd** [PonyGE/PonyGE2#89]
It looks to me that these deep_copy() calls are superfluous -- some crossover operators may need them (to avoid changing existing individuals), but others don't. I guess that they may be slow. So maybe the responsibility to avoid changing the input parents should be on the crossover operator itself, not here. 

https://github.com/PonyGE/PonyGE2/blob/40355f668757c8daa2503301c8c0eda54d566945/src/operators/crossover.py#L55
https://github.com/PonyGE/PonyGE2/blob/40355f668757c8daa2503301c8c0eda54d566945/src/operators/crossover.py#L56

**Comment by mikefenton** [comment]
When we do crossover on individuals we select pairs of individuals at random from the original parent pool. Since we can select the same individual multiple times, we have to create a new copy of them each time we select them from the original pool, as otherwise the original individual will be changed in-place and the next time they are selected for crossover from the pool they will be the changed version, not the original version.

**Comment by jmmcd** [comment]
My point is that only *some* crossover operators make changes to their inputs in-place.

**Comment by mikefenton** [comment]
Right, but the main crossover operator doesn't take this in to account, it just randomly selects pairs of individuals until the child pool is full. We don't know what the individual crossover functions are going to do, so to avoid any unwanted effects (i.e. people writing their own crossover functions and getting unexpected results) we do it once where you have indicated.

**Comment by mikefenton** [comment]
We can do it in the individual crossover functions, but we would need to add it in to the docs that any new functions may need to also do this to avoid over-writing parents as operations on individuals are applied inplace.

**Comment by jmmcd** [comment]
Yeah that's what I was thinking of. But yeah the disadvantage is people could then shoot themselves in the foot.

**Comment by jmmcd** [comment]
After working with it a bit more, I now understand that the deep_copy() is not doing so much work except in the case of derivation-tree genomes. And I was able work around another problem that the deep_copy() was causing me. So, I'm happy to drop this now -- thanks.

**Comment by jmmcd** [comment]
Hmmm... does the same reasoning apply to mutation also? I guess it might not, if mutation is only ever applied to individuals which have just been created by crossover. That's a property of many GP algorithms, but not all. In some algorithms, some new individuals are created by mutation of an individual in the current population. At the moment there's no deep_copy() for mutation.
